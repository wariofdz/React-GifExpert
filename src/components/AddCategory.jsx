import { useState } from "react";

//setCategory, lo cambiamos por onNewCategory

export const AddCategory = ({onNewCategory}) => {

    const [inputValue, setInputValue] = useState('');

    const onInputChange = (event) => {
        //console.log(event.target.value);
        // también podemos hacer una desestructuración del elemento event recibiendo en una variable {target} y acceder a su contenido target.value.
        // Podriamos obviar el (event) porque recogemos el primer argumento que recibimos 
        setInputValue(event.target.value);
    };

    const onSubmit = (event) => {
        //console.log(event);
        event.preventDefault(); // No permite que haga la función por defecto, en este caso, no permite que refresque el formulario cuando hacemos enter.
        //console.log(inputValue);
        // Con este if evitamos que den al enter sin introducir nada o solo un caracter. Con el return, salimos de la función, evitando que introduzca
        // valores no deseados.
        if (inputValue.trim().length <= 1) return;

        // La función del useState nos permite mandarle un callback. Contendrá los elementos anteriores y hay que conservarlos junto con el nuevo elemento.
        //setCategory( categories => [inputValue, ...categories]);
        onNewCategory(inputValue.trim());
        setInputValue(''); // Limpiamos nuestro input una vez se ha presionado el enter, para adicionar una nueva categoria.
    }

    return (
        <form onSubmit={(event) => {onSubmit(event)}}>

            <input 
                type="text" 
                placeholder="Buscar Gifs" 
                value={inputValue} 
                onChange={(event) => {onInputChange(event)}}/>

        </form>
    );
};
