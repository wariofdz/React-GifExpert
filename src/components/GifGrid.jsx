//import { useEffect, useState } from "react";
//import { getGifs } from "../helpers/getGifs";
import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifItem } from "./GifItem";

// Es una buena practica, porque no vuelve a asignar un nuevo espacio en memoria cuando el componente se vuelve a redibujar.
// Pero no evita que cada vez que haya un cambio en nuestro estado, no se vaya a disparar de nuevo. Nunca debemos colocar la ejecución de una función
// dentro de un functional component, ya que cada vez que se renderize de nuevo el FC, se va a realizar la petición HTTP y no queremos hacerlo.
// Una vez recibamos la categoria, solo deberiamos ejecutarlo una unica vez. No importa si hay redibujos de React.
// Lo extraeremos de aqui y lo colocaremos en otro archivo externo. 
/*
const getGifs = async(category) => {

  const url = `https://api.giphy.com/v1/gifs/search?api_key=gMmsTL7boqkQOPy81nSDmD2yKy9x90Pv&q=${category}&limit=20`;
  const resp = await fetch(url);
  // data = [] --> nos aseguramos que la variable data será un array.
  const {data = []} = await resp.json();
  // Es lo mismo literalmente que tenemos en el archivo GifExpertApp.jsx, en category.map(...)
  const gifs = data.map( img => ({
      id: img.id,
      title: img.title,
      url: img.images.downsized_medium.url
  }));
  //console.log(gifs);
  return gifs;
}

Trasladamos la función a nueva carpeta - archivo.
*/

export const GifGrid = ({category}) => {
  
    // Custom Hook
    const { images, isLoading } = useFetchGifs(category);
    
    // const [images, setImages] = useState([]);

    // const getImages = async() => {
    //     /* La función getGifs te devuelve una promesa ya que es una función asincrona. */
    //     const newImages = await getGifs(category);
    //     setImages(newImages);
    // }

    // /* useEffect: Realiza efectos secundarios en nuestro componente. 
    //   El primer parametro es una función de CallBack. En este caso ejecuta la API de Giphy.
    //   El segundo parámetro [] provoca que si dejamos las dependencias vacias, solo se dispara 
    //   la primera vez que se construye nuestro componente. No se vuelven a realizar las peticiones 
    //   aunque nuestro componente se renderize de nuevo. */
    // useEffect( () => {    
    //     //getGifs(category);
    //     /* No es una buena practica crear una función asincrona dentro del useEffect ya que devuelve
    //        un error. Este Hook necesita que se devuelva siempre una función. Lo que si se puede hacer
    //        es llamar a una función externa a éste que sea asincrona. */
    //     getImages();
    // }, []);

    return (
      <>
          <h3>{category}</h3>
          {
            // El null no se renderiza
              isLoading ? (<h2>Cargando...</h2>) : null
              // También se puede hacer como isLoading && (<h2>Cargando...</h2>), es un AND lógico, cuando se ejecute 
              // como true el isLoading, carga también la segunda parte.
          }          
          <div className="card-grid">
          { 
            //images.map( ({id, title, url}) => ( --> desestructurada
            images.map ((image) => (
              <GifItem 
                  key={image.id}
                  // Existe otra forma de poder mandar los datos, con el operador spread (...image) para esparcir el contenido. De esta manera, el componente recibe cada una de las propiedades 
                  // de la imagen como propertys (props). Se usa cuando uno tiene muchas propiedades.
                  { ...image}
                  //title={image.title}
                  //url={image.url} 
                  />

              ))
          }
          </div>       
      </>
    );
};
