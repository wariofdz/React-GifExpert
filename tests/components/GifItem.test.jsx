// Me faltÃ³ agregar el import de la testing library react, para la variable render.
import { render, screen } from "@testing-library/react";
import { GifItem } from "../../src/components/GifItem";

describe('Test in component GifItem', () => { 
    
    // Faltaba crear las variables que contuviesen la info para poder hacer la prueba.
    const title = 'Dragon Ball';
    const url   = 'https://one-punch.com/saitama';

    test('should have match with snapshot', () => { 
        
        const {container} = render(<GifItem title={title} url={url} />);
        expect(container).toMatchSnapshot();

    });

    test('should to show the image with the URL and ALT indicated', () => { 
    
        render(<GifItem title={title} url={url} />);
        //screen.debug();        
        // expect(screen.getByRole('img').src).toBe(url);
        // expect(screen.getByRole('img').alt).toBe(title);
        const {src, alt} = screen.getByRole('img');
        expect(src).toBe(url);
        expect(alt).toBe(title);

    });

    test('should to show the title in the component', () => { 
    
        render(<GifItem title={title} url={url} />);
        expect(screen.getByText(title)).toBeTruthy();

    });

 });
