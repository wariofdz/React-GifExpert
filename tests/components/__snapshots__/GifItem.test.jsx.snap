// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in component GifItem should have match with snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Dragon Ball"
      src="https://one-punch.com/saitama"
    />
    <p>
      Dragon Ball
    </p>
  </div>
</div>
`;
